File: apps/android/app/src/main/java/com/pawwalk/android/ui/screens/AssistantViewModel.kt
package com.pawwalk.android.ui.screens

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.pawwalk.android.data.AssistantChatRequest
import com.pawwalk.android.data.Network
import com.pawwalk.android.data.Walker
import com.pawwalk.android.data.WalkerRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Drives the AI booking assistant screen over POST /assistant/chat (mirrors the
 * iOS AssistantViewModel — see docs/FUNCTIONAL-REVIEW.md N6). Resolves suggested
 * walker IDs to real records so the UI can offer "book this walker".
 */
class AssistantViewModel : ViewModel() {
    data class Message(val fromUser: Boolean, val text: String, val walkers: List<Walker> = emptyList())

    private val _messages = MutableStateFlow(
        listOf(
            Message(
                false,
                "Hi! Tell me where and when you need a walk — e.g. \"a walker in the Mission for my husky tomorrow at 3pm\".",
            )
        )
    )
    val messages: StateFlow<List<Message>> = _messages.asStateFlow()

    private val _sending = MutableStateFlow(false)
    val sending: StateFlow<Boolean> = _sending.asStateFlow()

    private var walkersById: Map<String, Walker> = emptyMap()

    fun send(text: String) {
        val trimmed = text.trim()
        if (trimmed.isEmpty() || _sending.value) return
        _messages.value = _messages.value + Message(fromUser = true, text = trimmed)
        viewModelScope.launch {
            _sending.value = true
            try {
                if (walkersById.isEmpty()) walkersById = WalkerRepository.fetchWalkers().associateBy { it.id }
                val reply = Network.api.assistantChat(AssistantChatRequest(trimmed))
                val walkers = reply.suggestedWalkers.mapNotNull { walkersById[it] }
                _messages.value = _messages.value + Message(false, reply.reply, walkers)
            } catch (e: Exception) {
                _messages.value = _messages.value + Message(false, "Sorry — I couldn't reach the assistant. Try again.")
            } finally {
                _sending.value = false
            }
        }
    }
}

File: apps/ios/PawWalk/Features/Assistant/AssistantViewModel.swift
import Foundation
import Observation

/// Drives the AI booking assistant screen. Calls `POST /assistant/chat`
/// (LangGraph + heuristic intent parser on the backend) and keeps a simple
/// transcript. Resolves suggested walker IDs to names via `/walkers`.
@MainActor
@Observable
final class AssistantViewModel {
    struct Message: Identifiable {
        let id = UUID()
        let fromUser: Bool
        let text: String
        /// Walkers the assistant suggested for this reply (resolved to real records).
        var walkers: [Walker] = []
    }

    private(set) var messages: [Message] = [
        Message(fromUser: false,
                text: "Hi! Tell me where and when you need a walk — e.g. \"a walker in the Mission for my husky tomorrow at 3pm\".")
    ]
    private(set) var isSending = false

    private var walkersByID: [String: Walker] = [:]

    func loadWalkers() async {
        guard walkersByID.isEmpty else { return }
        let walkers = (try? await APIClient.shared.walkers()) ?? []
        walkersByID = Dictionary(uniqueKeysWithValues: walkers.map { ($0.id, $0) })
    }

    func send(_ text: String) async {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, !isSending else { return }
        messages.append(Message(fromUser: true, text: trimmed))
        isSending = true
        defer { isSending = false }
        do {
            let reply = try await APIClient.shared.assistantChat(message: trimmed)
            let walkers = reply.suggestedWalkers.compactMap { walkersByID[$0] }
            messages.append(Message(fromUser: false, text: reply.reply, walkers: walkers))
        } catch {
            messages.append(Message(fromUser: false,
                                    text: "Sorry — I couldn't reach the assistant. Try again."))
        }
    }
}
